File: coco/src/test/java/coco/ep/M1Test.java
package coco.ep;
import org.junit.Assert;
import org.junit.Test;

public class M1Test {

    public interface TestTemplate<FT> extends coco.ep.m1.Factory<FT>, M0Test.TestTemplate<FT> {
        default void test() {
            M0Test.TestTemplate.super.test();

            Exp<FT> expr1 = this.sub(this.lit(1.0), this.lit(2.0));
            org.junit.Assert.assertEquals(-1.0, convert(expr1).eval(), 0.0);

            Exp<FT> expr2 = this.lit(2.0);
            org.junit.Assert.assertEquals(2.0, convert(expr2).eval(), 0.0);
            Assert.assertEquals(convert(this.sub(this.lit(1.0), this.lit(2.0))).eval(), -1.0, 0.0);
        }
    }
    private static class ActualTest implements M1Test.TestTemplate<coco.ep.m0.finalized.Exp>, coco.ep.m1.finalized.Factory {}

    @Test
    public void testTest() { new M1Test.ActualTest().test(); }
}

File: coco_scala/src/test/scala/scans/m1/M1Test.scala
package scans.m1

import org.scalatest.FunSuite

trait TestTemplate[FT] extends Factory[FT] {
  val suite:FunSuite
  import suite._

  def test() : Unit = {

    val circ1 = convert(this.identity(1))
    val circ2 = convert(this.identity(1))
    val circ3 = convert(this.identity(3))
    assert(circ1.width == circ2.width)
    assert(circ1.width != circ3.width)

    val circ4 = convert(api.brentKung(this))
    assert(circ4.width == 4)
  }
}

class M0Test extends FunSuite { self =>

  object ActualTest extends TestTemplate[finalized.Circuit] with finalized.Factory {
    val suite: FunSuite = self
  }

  test("M0") {
    ActualTest.test()
  }
}
